File: src/clase-06/typescript/observer.ts
interface Subscriber {
  update(data: any): void;
}

class Publisher {
  private subscribers: Subscriber[] = [];

  subscribe(s: Subscriber) {
    this.subscribers.push(s);
  }
  unsubscribe(s: Subscriber) {
    this.subscribers = this.subscribers.filter(x => x !== s);
  }
  protected notify(data: any) {
    this.subscribers.forEach(s => s.update(data));
  }
}

class Stock extends Publisher {
  private price = 0;

  setPrice(p: number) {
    this.price = p;
    this.notify(this.price);
  }
}

class EmailAlert implements Subscriber {
  update(data: any): void {
    console.log(`[Email] Nuevo precio: ${data}`);
  }
}

class Dashboard implements Subscriber {
  update(data: any): void {
    console.log(`[Dashboard] Precio actualizado: ${data}`);
  }
}

// Uso
const stock = new Stock();
const e = new EmailAlert();
const d = new Dashboard();
stock.subscribe(e);
stock.subscribe(d);
stock.setPrice(200.75);

File: src/clase-06/python/observer.py
from abc import ABC, abstractmethod

class Subscriber(ABC):
    @abstractmethod
    def update(self, data): ...

class Publisher:
    def __init__(self):
        self._subscribers = []

    def subscribe(self, sub: Subscriber):
        self._subscribers.append(sub)

    def unsubscribe(self, sub: Subscriber):
        self._subscribers.remove(sub)

    def notify(self, data):
        for sub in list(self._subscribers):
            sub.update(data)

class Stock(Publisher):
    def __init__(self):
        super().__init__()
        self._price = 0

    def set_price(self, p: float):
        self._price = p
        self.notify(self._price)

class EmailAlert(Subscriber):
    def update(self, price):
        print(f"[Email] Nuevo precio: {price}")

class MobileDisplay(Subscriber):
    def update(self, price):
        print(f"[App] Precio actualizado: {price}")

# Uso
stock = Stock()
ea = EmailAlert()
md = MobileDisplay()

stock.subscribe(ea)
stock.subscribe(md)
stock.set_price(100.5)
